### PR TITLE
Make component nudging work for gitlab repositories

### PR DESCRIPTION
**Pull request by skabashnyuk**
### What does this PR do?
 Make component nudging work for GitLab repositories

We have two types of Renovate updates. That is Tekton task updates and Component docker image dependency nudging. In both cases, we are first looking for a list of `UpdateTarget`. That is the list of source code repositories that have to be updated with some extra information. After that, we transform that list of targets to the structure `AuthenticatedUpdateConfig` that describes what we want Renovate to do with that concrete target.  At the end, this update configuration is executed on the k8s cluster.

```mermaid
sequenceDiagram
    participant ComponentDependenciesUpdater
    participant UpdateTargetProvider
    participant UpdateCoordinator

    ComponentDependenciesUpdater->>UpdateTargetProvider: What should we update?
    UpdateTargetProvider-->>ComponentDependenciesUpdater:  list of UpdateTarget
    loop ComponentDependenciesUpdater
        ComponentDependenciesUpdater->>ComponentDependenciesUpdater: For each target create Renovate's update config
    end
    ComponentDependenciesUpdater-->>UpdateCoordinator: Execute update
```

1.  Component docker image dependency nudging.  In this case `UpdateTarget` transformed to `AuthenticatedUpdateConfig` with `NewNudgeDependencyUpdateConfig` method. Execution implemented as `PipelineRun` on user's namespace.
2.  Tekton task updates.  In this case `UpdateTarget` transformed to `AuthenticatedUpdateConfig` with `NewTektonUpdateConfig` method. Execution implemented as k8s `Job` on build-service namespace.
 

### Screenshot/screencast of this PR
Знімок екрана 2024-06-05 о 11 58 09
Знімок екрана 2024-06-05 о 11 58 02

Знімок екрана 2024-06-05 о 11 56 10
Знімок екрана 2024-06-05 о 11 50 44

### What issues does this PR fix or reference?

Context: https://issues.redhat.com/browse/STONEBLD-2317

### How to test this PR?
 
1. Add Gitlab secret into Openshift

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: pipelines-as-code-secret
  namespace: <YOUR NAMESPACE>
  labels:
    appstudio.redhat.com/credentials: scm
    appstudio.redhat.com/scm.host: <your-gitlab-host> # eg. gitlab.cee.redhat.com 
type: kubernetes.io/basic-auth
stringData:
  password: <YOUR GL ACCESS TOKEN>
```
2. Create a parent component  from repo  https://github.com/skabashnyuk/common-nodejs-parent
```yaml
apiVersion: appstudio.redhat.com/v1alpha1
kind: Component
metadata:
  name: ${COMPONENT_NAME_NUDGE_PARENT}
  annotations:
    build.appstudio.openshift.io/request: configure-pac
    image.redhat.com/generate: '{"visibility": "public"}'
    git-provider:  ${GIT_PROVIDER}
    git-provider-url: ${GIT_PROVIDER_URL}
spec:
  application: ${COMPONENT_APPLICATION_NAME}
  build-nudges-ref:
    - ${COMPONENT_BUILD_NUDGES_REF}
  componentName: ${COMPONENT_NAME_NUDGE_PARENT}
  replicas: 0
  resources:
    limits:
      cpu: 15m
      memory: 30Mi
    requests:
      cpu: 10m
      memory: 25Mi
  source:
    git:
      context: ./
      devfileUrl: ${DEVFILE_URL}
      dockerfileUrl: ${COMPONENT_NAME_NUDGE_PARENT_DOCKERFILE_URL}
      revision: main
      url: ${COMPONENT_NAME_NUDGE_PARENT_URL}
```
3. Create component from https://github.com/skabashnyuk/node-express-hello
```yaml
apiVersion: appstudio.redhat.com/v1alpha1
kind: Component
metadata:
  name: ${COMPONENT_NAME}
  annotations:
    build.appstudio.openshift.io/request: configure-pac
    image.redhat.com/generate: '{"visibility": "public"}'
    git-provider:  ${GIT_PROVIDER}
    git-provider-url: ${GIT_PROVIDER_URL}
spec:
  application: ${COMPONENT_APPLICATION_NAME}
  componentName: ${COMPONENT_NAME}
  replicas: 0
  resources:
    limits:
      cpu: 15m
      memory: 30Mi
    requests:
      cpu: 10m
      memory: 25Mi
  source:
    git:
      context: ./
      devfileUrl: ${DEVFILE_URL}
      dockerfileUrl: ${COMPONENT_DOCKERFILE_URL}
      revision: main
      url: ${COMPONENT_URL}
```

### CODE REVIEW

File: config/rbac/role.yaml [hunk 1]
```diff
@@ -2,6 +2,7 @@
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
+  creationTimestamp: null
```
Comment (mmorhun) [inline config/rbac/role.yaml:5]: What's the purpose of the change?

File: controllers/component_dependency_update_controller.go [hunk 2]
```diff
@@ -79,20 +80,23 @@ var delayTime = time.Second * 10
 
 // ComponentDependencyUpdateReconciler reconciles a PipelineRun object
 type ComponentDependencyUpdateReconciler struct {
-	client.Client
-	ApiReader      client.Reader
-	Scheme         *runtime.Scheme
-	EventRecorder  record.EventRecorder
-	UpdateFunction UpdateComponentDependenciesFunction
+	apiReader                    client.Reader
+	client                       client.Client
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller.go:84]: Usually these fields are public. Is there any reason to make them private?

File: controllers/component_dependency_update_controller.go [hunk 3]
```diff
@@ -339,23 +339,42 @@ func (r *ComponentDependencyUpdateReconciler) handleCompletedBuild(ctx context.C
 	}
 
 	for i := range components.Items {
-		comp := components.Items[i]
-		if slices.Contains(updatedComponent.Spec.BuildNudgesRef, comp.Name) {
-			toUpdate = append(toUpdate, comp)
+		component := components.Items[i]
+		if slices.Contains(updatedComponent.Spec.BuildNudgesRef, component.Name) {
+			gitProvider, err := getGitProvider(component)
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller.go:344]: `getGitProvider` is and expensive operation. Can we reduce number of its calls in a reasonable way?

File: controllers/component_dependency_update_controller.go [hunk 4]
```diff
@@ -339,23 +339,42 @@ func (r *ComponentDependencyUpdateReconciler) handleCompletedBuild(ctx context.C
...
+				// component misconfiguration shouldn't prevent other components from being updated
+				// deepcopy the component to avoid implicit memory aliasing in for loop
+				r.eventRecorder.Event(component.DeepCopy(), "Warning", "ErrorComponentProviderInfo", err.Error())
+				continue
+			}
+			scmComponent, err := git.NewScmComponent(gitProvider, component.Spec.Source.GitSource.URL, component.Spec.Source.GitSource.Revision, component.Name, component.Namespace)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			toUpdate = append(toUpdate, scmComponent)
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller.go:355]: I'd rename to `componentsToUpdate` or `scmComponentsToUpdate`

File: controllers/component_dependency_update_controller_test.go [hunk 5]
```diff
@@ -50,25 +52,46 @@ const (
 	ImageDigest   = "IMAGE_DIGEST"
 )
 
-var failures = 0
+//var failures = 0
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller_test.go:55]: Leftover?

File: controllers/component_dependency_update_controller_test.go [hunk 6]
```diff
@@ -50,25 +52,46 @@ const (
...
-		//TODO: do we want the event? It's just for unit testing at the moment
-		eventRecorder.Event(&downstreamComponent, v1.EventTypeNormal, ComponentNudgedEventType, fmt.Sprintf("component %s.%s was nudged by successful build of %s that produces image %s:%s@%s", downstreamComponent.Namespace, downstreamComponent.Name, result.Component.Name, result.BuiltImageRepository, result.BuiltImageTag, result.Digest))
+func (t *TestComponentDependenciesUpdater) Update(ctx context.Context, components []*git.ScmComponent, buildResult *renovate.BuildResult) error {
+	if t.failures == 0 {
+		for _, downstreamComponent := range components {
+			log.Info("Nudging due to successful build",
+				"component", downstreamComponent.ComponentName(),
+				"updatedcomponent", buildResult.UpdatedComponentName,
+				l.Action, l.ActionUpdate)
+			//TODO: do we want the event? It's just for unit testing at the moment
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller_test.go:78]: This comment is detached from it's original code.
Do we really need this event now?

File: controllers/component_dependency_update_controller_test.go [hunk 7]
```diff
@@ -50,25 +52,46 @@ const (
...
+			t.eventRecorder.Event(component, v1.EventTypeNormal, ComponentNudgedEventType, fmt.Sprintf("component %s.%s was nudged by successful build of %s that produces image %s:%s@%s", downstreamComponent.NamespaceName(), downstreamComponent.ComponentName(), buildResult.UpdatedComponentName, buildResult.BuiltImageRepository, buildResult.BuiltImageTag, buildResult.Digest))
 
+		}
+		if err := t.componentDependenciesUpdater.Update(ctx, components, buildResult); err != nil {
+			return err
+		}
+		return nil
 	}
-	return false, nil
+	t.failures = t.failures - 1
```
Comment (mmorhun) [inline controllers/component_dependency_update_controller_test.go:93]: Why do we decrease failures count? Or is it a remainder? Please make it clear via name or comment.

File: controllers/suite_test.go [hunk 8]
```diff
@@ -52,11 +52,12 @@ import (
...
-	testEnv   *envtest.Environment
-	ctx       context.Context
-	cancel    context.CancelFunc
-	log       logr.Logger
+	k8sClient           client.Client
+	testEnv             *envtest.Environment
+	ctx                 context.Context
+	cancel              context.CancelFunc
+	log                 logr.Logger
+	dependenciesUpdater *TestComponentDependenciesUpdater
```
Comment (mmorhun) [inline controllers/suite_test.go:60]: Values here are global for all tests. please move controller specific stuff to the controller itself.

File: controllers/suite_test.go [hunk 9]
```diff
@@ -155,13 +156,9 @@ var _ = BeforeSuite(func() {
...
-	err = (&ComponentDependencyUpdateReconciler{
-		Client:         k8sManager.GetClient(),
-		ApiReader:      k8sManager.GetAPIReader(),
-		Scheme:         k8sManager.GetScheme(),
-		EventRecorder:  k8sManager.GetEventRecorderFor("ComponentDependencyUpdateReconciler"),
-		UpdateFunction: failingDependencyUpdate,
-	}).SetupWithManager(k8sManager)
+	componentDependencyUpdateReconcilerEventRecorder := k8sManager.GetEventRecorderFor("ComponentDependencyUpdateReconciler")
+	dependenciesUpdater = NewTestComponentDependenciesUpdater(k8sManager.GetClient(), k8sManager.GetScheme(), componentDependencyUpdateReconcilerEventRecorder)
+	err = NewComponentDependencyUpdateReconciler(k8sManager.GetClient(), k8sManager.GetAPIReader(), componentDependencyUpdateReconcilerEventRecorder, dependenciesUpdater).SetupWithManager(k8sManager)
```
Comment (mmorhun) [inline controllers/suite_test.go:161]: Please format accordingly

File: main.go [hunk 10]
```diff
@@ -184,13 +184,7 @@ func main() {
...
 	}
 
-	if err = (&controllers.ComponentDependencyUpdateReconciler{
-		Client:         mgr.GetClient(),
-		ApiReader:      mgr.GetAPIReader(),
-		Scheme:         mgr.GetScheme(),
-		EventRecorder:  mgr.GetEventRecorderFor("ComponentDependencyUpdateReconciler"),
-		UpdateFunction: controllers.DefaultDependenciesUpdate,
-	}).SetupWithManager(mgr); err != nil {
+	if err = (controllers.NewDefaultComponentDependencyUpdateReconciler(mgr.GetClient(), mgr.GetAPIReader(), mgr.GetScheme(), mgr.GetEventRecorderFor("ComponentDependencyUpdateReconciler"))).SetupWithManager(mgr); err != nil {
```
Comment (mmorhun) [inline main.go:187]: Please format accordingly to the above

File: pkg/renovate/nudge.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+	updateTargets := d.updateTargetProvider.GetUpdateTargets(ctx, components)
+	log.V(l.DebugLevel).Info("found renovate targets", "targets", len(updateTargets))
+	if err := d.updateCoordinator.ExecutePipelineRun(ctx, buildResult.UpdatedComponentNamespace, UpdateTargetToNudgeUpdate(updateTargets, buildResult)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func NewNudgeDependencyUpdateConfig(buildResult *BuildResult, platform, endpoint, username, gitAuthor string, repositories []*Repository) UpdateConfig {
+
```
Comment (mmorhun) [inline pkg/renovate/nudge.go:60]: extra new line

File: pkg/renovate/nudge.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+						DatasourceTemplate:   "docker",
+						CurrentValueTemplate: buildResult.BuiltImageTag,
+						DepNameTemplate:      buildResult.BuiltImageRepository,
+					},
+				},
+			},
+		},
+		Endpoint:            endpoint,
+		Repositories:        repositories,
+		ForkProcessing:      "enabled",
```
Comment (mmorhun) [inline pkg/renovate/nudge.go:106]: it might be out of scope of this PR, but do we really need this option?

File: pkg/renovate/update.go [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+		return err
+	}
+	if err := j.client.Update(ctx, configMap); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (j *UpdateCoordinator) ExecuteUpdateWithLimits(ctx context.Context, updates []*AuthenticatedUpdateConfig) error {
+
```
Comment (mmorhun) [inline pkg/renovate/update.go:195]: extra new line

File: pkg/renovate/update.go [hunk 14]
```diff
@@ -0,0 +1,342 @@
...
+		if end > len(updates) {
+			end = len(updates)
+		}
+		err := j.ExecuteUpdate(ctx, updates[i:end])
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+
```
Comment (mmorhun) [inline pkg/renovate/update.go:208]: extra new line

**Comment by rcerven** [comment]
closed in favor of https://github.com/konflux-ci/build-service/pull/347
